### PR TITLE
Generate: fix speculative decoding

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR:
- Fixes speculative decoding quality:
  - Incorrect indexing operation
  - The assistant model should sample when the larger model also samples (more generally, it should take the original model's `generation_config`)
  - Custom logits processors should also be passed to the assistant model
- Changes docs to put an emphasis on "speculative decoding" as opposed to "assisted generation", as the former is more popular

________

`RUN_SLOW=1 py.test tests/ -k speculative` was run locally to confirm that slow assisted generation whisper tests were passing.

### CODE REVIEW

File: src/transformers/generation/candidate_generator.py [hunk 1]
```diff
@@ -82,31 +85,20 @@ class AssistedCandidateGenerator(CandidateGenerator):
...
         logits_processor: "LogitsProcessorList",
         model_kwargs: Dict,
         inputs_tensor: Optional[torch.Tensor] = None,
-        eos_token_id: Optional[Union[int, List[int]]] = None,
     ):
+        # Prepare the assistant and the starting number of candidate tokens
         self.assistant_model = assistant_model
-
-        # Prepare the number of candidate tokens
-        if hasattr(assistant_model, "num_assistant_tokens"):
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:101]: removed as the next release will be v4.37

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1673,7 +1673,7 @@ def generate(
             )
 
         # 8. prepare distribution pre_processing samplers
-        logits_processor = self._get_logits_processor(
+        prepared_logits_processor = self._get_logits_processor(
```
Comment (gante) [inline src/transformers/generation/utils.py:1676]: changed the name so we can keep a separation between the custom logits processors (`logits_processor`) and the custom + loaded logits processors (now `prepared_logits_processor`)

assisted generation should know which logits processors are custom

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -4789,17 +4789,9 @@ def _speculative_sampling(
     # Gets the probabilities from the logits. q_i and p_i denote the assistant and model probabilities of the tokens
     # selected by the assistant, respectively.
     q = candidate_logits.softmax(dim=-1)
-    q_i = q[
-        :,
-        torch.range(0, candidate_length - 1, dtype=torch.int),
-        candidate_input_ids[:, -candidate_length:],
-    ].squeeze(0, 1)
+    q_i = q[:, torch.arange(candidate_length), candidate_input_ids[:, -candidate_length:]].squeeze(0, 1)
```
Comment (gante) [inline src/transformers/generation/utils.py:4792]: torch.range is deprecated 🙃

File: docs/source/en/generation_strategies.md [hunk 4]
```diff
@@ -339,13 +339,16 @@ This guide illustrates the main parameters that enable various decoding strategi
...
-### Assisted Decoding
+### Speculative Decoding
 
-Assisted decoding is a modification of the decoding strategies above that uses an assistant model with the same
-tokenizer (ideally a much smaller model) to greedily generate a few candidate tokens. The main model then validates
-the candidate tokens in a single forward pass, which speeds up the decoding process. Currently, only greedy search
-and sampling are supported with assisted decoding, and doesn't support batched inputs. To learn more about assisted
-decoding, check [this blog post](https://huggingface.co/blog/assisted-generation).
+Speculative decoding (also known as assisted decoding) is a modification of the decoding strategies above that uses an
+assistant model with  the same tokenizer, ideally a much smaller model, to generate a few candidate tokens. The main
```
Comment (amyeroberts) [inline docs/source/en/generation_strategies.md:345]: ultra-nit 

```suggestion
assistant model with the same tokenizer, ideally a much smaller model, to generate a few candidate tokens. The main
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

I think you know I'm going to ask about tests 😄 Just to confirm that `RUN_SLOW=1 py.test tests/ -k speculative` was failing before and is now passing with this change? If it was passing before, can we add a test which would have caught the failure?

**Comment by gante** [comment]
@amyeroberts all tests were passing, the quality of the outputs was meh in some cases. 

Going to add a slow test with a parameterization corresponding to a bad output in the previous commit, so we don't regress :)

File: docs/source/en/generation_strategies.md [hunk 5]
```diff
@@ -367,7 +370,7 @@ To enable assisted decoding, set the `assistant_model` argument with a model.
 ```
 
 When using assisted decoding with sampling methods, you can use the `temperature` argument to control the randomness
-just like in multinomial sampling. However, in assisted decoding, reducing the temperature will help improving latency.
+just like in multinomial sampling. However, in assisted decoding, reducing the temperature may help improving latency.
```
Comment (merveenoyan) [inline docs/source/en/generation_strategies.md:373]:
```suggestion
just like in multinomial sampling. However, in assisted decoding, reducing the temperature may help improve the latency.
```

**Review by merveenoyan — COMMENTED** [review]
Have made some suggestions for clarity 😊
